lib/common/common/src/top_k.rs

```diff
@@ -34,27 +34,37 @@ impl TopK {
 
     /// Returns the minimum score of the top k elements.
     ///
-    /// Updated every 2k elements.
+    /// Updated every 2k elements unless forced
     /// Initially set to `ScoreType::MIN`.
     pub fn threshold(&self) -> ScoreType {
         self.threshold
     }
 
+    // Force updates the threshold
+    pub fn update_threshold(&mut self) {
+        let position_to_sort = self.k.min(self.elements.len());
+        let (_, sorted_el, _) = self.elements.select_nth_unstable(position_to_sort - 1);
+        self.threshold = sorted_el.0.score;
+        self.elements.truncate(position_to_sort);
+    }
+
     pub fn push(&mut self, element: ScoredPointOffset) {
         if element.score > self.threshold {
             self.elements.push(Reverse(element));
             // check if full
             if self.elements.len() == self.k * 2 {
-                let (_, median_el, _) = self.elements.select_nth_unstable(self.k - 1);
-                self.threshold = median_el.0.score;
-                self.elements.truncate(self.k);
+                self.update_threshold();
             }
         }
     }
 
-    pub fn into_vec(mut self) -> Vec<ScoredPointOffset> {
+    pub fn truncate(&mut self) {
         self.elements.sort_unstable();
         self.elements.truncate(self.k);
+    }
+
+    pub fn into_vec(mut self) -> Vec<ScoredPointOffset> {
+        self.truncate();
         self.elements.into_iter().map(|Reverse(x)| x).collect()
```

Comment:
I am not sure if compiler is able to optimize this, but just to be extra sure, I would prefer to avoid using Reverse. There is a `select_nth_unstable_by` function, which should help

Reply:
Done